File: a1/A1/src/a1/Matrix.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package a1;

/**
 *
 * @author jssmythe
 */
public class Matrix {
    
    public int cols;
    public int rows;
    
    public double[][] matrix;
    
    public Matrix() {
        this.cols = 4;
        this.rows = 4;
        
        this.matrix = new double[][] {{1,0,0,0},{0,1,0,0},{0,0,1,0},{0,0,0,1}};
        
    }
    
    public Matrix(int rows, int cols, double[][] array) {
        this.cols = cols;
        this.rows = rows;
        
        this.matrix = array;
    }
    
    public Matrix transpose() {
        
        double[][] transp = new double[cols][rows];
        
        for(int i=0;i<rows;i++) {
            for(int j=0;j<cols;j++) {
                transp[i][j] = this.matrix[j][i];
            }
        }
        
        Matrix result = new Matrix(cols,rows,transp);
        
        return result;
    }
    
    public boolean equality(Matrix neo) {
        
        if((this.cols == neo.cols) && (this.rows == neo.rows)) {
            for(int i=0;i<rows;i++) {
                for(int j=0;j<cols;j++) {
                    if(this.matrix[i][j] != neo.matrix[i][j]) {
                        return false;
                    }
                }
            }
            return true;
        }
        
        return false;
    }
    
    public Matrix multiplication(Matrix neo) {
        
        if(this.cols != neo.rows) {
            return null;
        }
        
        double[][] multi = new double[this.rows][neo.cols];
        
        for(int i=0;i<this.rows;i++) {
            for(int j=0;j<neo.cols;j++) {
                int sum = 0;
                for(int k=0;k<this.cols;k++) {
                    sum += this.matrix[i][k] * neo.matrix[k][j];
                }
                multi[i][j] = (int)sum;
            }
        }
        
        Matrix result = new Matrix(this.rows, neo.cols, multi);
        
        return result;
    }
    
    public Matrix inverse() {
        
        return this;
    }
    
    public int dotProduct(Matrix neo) {
        
        if(this.cols != 1 || neo.cols != 1 || this.rows != neo.rows) {
            System.out.println("Matrices are not Vectors Matrices, Dot Product requires two Nx1 matrices");
            return Integer.MAX_VALUE;
        }
        
        int result = 0;
        
        for(int i=0;i<this.rows;i++) {
            result += this.matrix[i][0] * neo.matrix[i][0];
        }
        
        return (int)result;
    }
    
    public Matrix crossProduct(Matrix neo) {
        
        if(this.cols != 1 || neo.cols != 1 || this.rows != 3 || neo.rows != 3) {
            System.out.println("Matrices are not Vectors Matrices, Cross Product requires two 3x1 matrices");
            return null;
        }
        
        double x = this.matrix[0][0];
        double y = this.matrix[1][0];
        double z = this.matrix[2][0];
        
        
        Matrix dual = new Matrix(this.rows, this.rows, new double[][] {{0,-z,y},{z,0,-x},{-y,x,0}});
        
        Matrix result = dual.multiplication(neo);
        
        return result;
    }
    
    public void showMatrix() {
        
        for(int i=0;i<this.rows;i++) {
            for(int j=0;j<this.cols;j++) {
                System.out.print(this.matrix[i][j]);
                System.out.print(" ");
            }
            System.out.print("\n");
        }
    }
    
}

File: a1/c++/A1/Matrix.cpp
/* 
 * File:   Matrix.cpp
 * Author: jssmythe
 * 
 * Created on January 27, 2014, 5:40 PM
 */

#include "Matrix.h"
#include <iostream>
#include <iomanip>

using namespace std;

Matrix::Matrix() {
}

Matrix::Matrix(int rows, int cols, double array[]) {
    
    this->rows = rows;
    this->cols = cols;
    
    this->matrix = new double*[rows];
    for(int i=0;i<rows;i++) {
        this->matrix[i] = new double[cols];
    }
    
    for(int i=0;i<rows;i++) {
        for(int j=0;j<cols;j++) {
            this->matrix[i][j] = array[j + (cols*i)];
        }
    }
}

Matrix* Matrix::transpose() {
    
    double transp[this->cols*this->rows];
        
    for(int i=0;i<this->rows;i++) {
        for(int j=0;j<this->cols;j++) {
            transp[j + (cols*i)] = this->matrix[j][i];
        }
    }

    return new Matrix(this->cols,this->rows,transp); 
}

bool Matrix::equality(Matrix* neo) {
    
    if((this->cols == neo->cols) && (this->rows == neo->rows)) {
        for(int i=0;i<this->rows;i++) {
            for(int j=0;j<this->cols;j++) {
                if(this->matrix[i][j] != neo->matrix[i][j]) {
                    return false;
                }
            }
        }
        return true;
    }
        
    return false;
    
}

Matrix* Matrix::multiplication(Matrix* neo) {
    
    int nCols = neo->cols;
    int nRows = neo->rows;
    double **nMatrix = neo->matrix;
    
    if(this->cols != nRows) {
        cout << "Matrix Size Mismatch" << endl;
        return NULL;
    }
        
    double multi[this->rows * this->cols];

    for(int i=0;i<this->rows;i++) {
        for(int j=0;j<nCols;j++) {
            int sum = 0;
            for(int k=0;k<this->cols;k++) {
                sum += this->matrix[i][k] * nMatrix[k][j];
            }
            multi[j + (nCols*i)] = (int)sum;
        }
    }

    return new Matrix(this->rows, nCols, multi);
    
}

Matrix* Matrix::inverse() {
    
}

int Matrix::dotProduct(Matrix* neo) {
    
    if(this->cols != 1 || neo->cols != 1 || this->rows != neo->rows) {
        cout << "Matrices are not Vectors Matrices, Dot Product requires two Nx1 matrices" << endl;
        return INT_MAX;
    }

    int result = 0;

    for(int i=0;i<this->rows;i++) {
        result += this->matrix[i][0] * neo->matrix[i][0];
    }

    return (int)result;
    
}

Matrix* Matrix::crossProduct(Matrix* neo) {
    
    int nRows = neo->rows;
    int nCols = neo->cols;
    
    if(this->cols != 1 || nCols != 1 || this->rows != 3 || nRows != 3) {
        cout << "Matrices are not Vectors Matrices, Cross Product requires two 3x1 matrices" << endl;
        return NULL;
    }

    double x = this->matrix[0][0];
    double y = this->matrix[1][0];
    double z = this->matrix[2][0];
    
    double temp[] = {0,-z,y,z,0,-x,-y,x,0};

    Matrix* dual = new Matrix(this->rows, this->rows, temp);
    
    double thing[3];

    return dual->multiplication(neo); 
}

void Matrix::showMatrix() {
    
    for(int i=0;i<this->rows;i++) {
        for(int j=0;j<this->cols;j++) {
            cout << setw(6) << setprecision(3) << this->matrix[i][j];
            
        }
        cout << endl;
    }
    cout << endl;
}

Matrix::~Matrix() {
}

